addons/t9n/static/src/core/project_list.js

```diff
@@ -0,0 +1,81 @@
+import { Component, onMounted, useState } from "@odoo/owl";
+
+import { formatList } from "@web/core/l10n/utils";
+import { useService } from "@web/core/utils/hooks";
+
+export class ProjectList extends Component {
+    static props = {};
+    static template = "t9n.ProjectList";
+
+    setup() {
+        this.action = useService("action");
+        this.state = useState({
+            filters: {
+                searchText: "",
+            },
+            sorting: {
+                column: "name",
+                order: "asc",
+            },
+        });
+        this.store = useState(useService("t9n.store"));
+        onMounted(() => {
+            this.store.fetchProjects();
+        });
+    }
+
+    formatTargetLanguages(langs) {
+        return formatList(
+            langs.map((lang) => {
+                return lang.name;
+            })
+        );
+    }
+
+    onClickProject(id) {
+        this.action.doAction({
+            type: "ir.actions.act_window",
+            res_id: id,
+            res_model: "t9n.project",
+            views: [[false, "form"]],
+            target: "new",
+        });
+    }
+
+    isFilteredOut(project) {
+        if (this.state.filters.searchText === "") {
+            return false;
+        }
+        return !project.name.includes(this.state.filters.searchText);
+    }
+
+    handleSort(column) {
+        if (this.state.sorting.column === column) {
+            this.state.sorting.order = this.state.sorting.order === "asc" ? "desc" : "asc";
+        } else {
+            this.state.sorting.column = column;
+            this.state.sorting.order = "asc";
+        }
+        this.applySortingLogic();
+    }
+
+    applySortingLogic() {
+        this.store.projects.sort((p1, p2) => {
```

Comment:
The way things are sorted is essentially a matter of display. You could imagine a situation where you have two different components side-by-side, displaying the same data, but in a different order. In that case, if they both changed the order of `this.store.projects`, it just wouldn't work. That's why you want to make a copy of the data and apply the sorting to that copy, not directly to the original data.

Reply:
Another, more concrete problem with your current implementation is that it's too imperative, and as such, too dependent on the flows: you have to explicitly call that function for the sort to be applied. And it's already responsible for a problem: the default sort (order by name, asc) is not applied when you first open the page, since you don't call it in the setup of the component. Easy fix you might think, just call `applySortingLogic` in the setup! But then, what if some random flow pushes another project into the list? You shouldn't forget to call `applySortingLogic` there either. And what if you change the name of a project in your list, and, no luck, it is used to sort your list? Ditto, don't forget to call `applySortingLogic`!

Can you see how imperative programming (i.e. explicitly calling functions, "do this, do that") becomes overwhelming really quickly when it comes to managing state? It's really easy to forget about one single flow somewhere on the other side of the application and end up with inconsistent state.

Fortunately, the solution exists; it's called "reactivity". Reactivity of Owl components allows you to re-render your component based solely on state: no need to call any function. To get rid of the function call, the trick is to make it so that every time your state changes and triggers a re-render, the sorted list is recomputed. My usual solution for this is to put the value you want to recompute inside a getter:
```js
get projects() {
    const searchTerms = this.state.filters.searchText.trim().toUpperCase();
    const projects = this.store.projects.filter((p) =>
        p.name.toUpperCase().includes(searchTerms)
    );
    projects.sort((p1, p2) => {
        // reimplement your sorting logic here
    });
    return projects;
}
```
```xml
<t t-foreach="this.projects" t-as="project" t-key="project.id">
```